File: Assets/Bundles/Shared/Resources/TS/UserInput/Preferred.lua
-- Compiled with unity-ts v2.1.0-75
local Bin = require("Shared/TS/Util/Bin").Bin
local Signal = require("Shared/TS/Util/Signal").Signal
local PreferredDriver = require("Shared/TS/UserInput/Drivers/PreferredDriver").PreferredDriver
-- * Utility class for observing the player's currently-used control scheme.
local Preferred
do
	Preferred = setmetatable({}, {
		__tostring = function()
			return "Preferred"
		end,
	})
	Preferred.__index = Preferred
	function Preferred.new(...)
		local self = setmetatable({}, Preferred)
		return self:constructor(...) or self
	end
	function Preferred:constructor()
		self.bin = Bin.new()
		self.preferredDriver = PreferredDriver:instance()
		self.ControlSchemeChanged = Signal.new()
		self.bin:Add(self.ControlSchemeChanged)
		self.controlScheme = self.preferredDriver:GetScheme()
		self.bin:Connect(self.preferredDriver.SchemeChanged, function(scheme)
			self.controlScheme = scheme
			self.ControlSchemeChanged:Fire(self.controlScheme)
		end)
	end
	function Preferred:GetControlScheme()
		return self.controlScheme
	end
	function Preferred:ObserveControlScheme(observer)
		local cleanup
		cleanup = observer(self.controlScheme)
		local disconnectChanged = self.ControlSchemeChanged:Connect(function(controlScheme)
			local _result = cleanup
			if _result ~= nil then
				_result()
			end
			cleanup = observer(controlScheme)
		end)
		local observing = true
		local stopObserving = function()
			if not observing then
				return nil
			end
			observing = false
			disconnectChanged()
			if cleanup ~= nil then
				cleanup()
				cleanup = nil
			end
		end
		self.bin:Add(stopObserving)
		return stopObserving
	end
	function Preferred:Destroy()
		self.bin:Destroy()
	end
end
return {
	Preferred = Preferred,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Bundles/@Easy/Core/Core~/src/Shared/UserInput/Preferred.ts
import { Bin } from "Shared/Util/Bin";
import { Signal } from "Shared/Util/Signal";
import { Game } from "../Game";

export type ControlScheme = "MouseKeyboard" | "Touch";

/** Utility class for observing the player's currently-used control scheme. */
export class Preferred {
	private readonly bin = new Bin();

	/** A signal that fires every time the currently-used control scheme changes. */
	public readonly controlSchemeChanged = new Signal<[controlScheme: ControlScheme]>();

	private controlScheme!: ControlScheme;

	constructor() {
		this.bin.Add(this.controlSchemeChanged);
		this.InitControlScheme();
	}

	private InitControlScheme(): void {
		let platform = Game.platform;
		if (platform === AirshipPlatform.Android || platform === AirshipPlatform.iOS) {
			this.controlScheme = "Touch";
		} else {
			this.controlScheme = "MouseKeyboard";
		}
	}

	/** Get the currently-used control scheme. */
	public GetControlScheme() {
		return this.controlScheme;
	}

	/**
	 * Observe the currently-used control scheme. Fires the `observer` function immediately,
	 * as well as every time the control scheme changes.
	 *
	 * The returned function can be called to stop the observation process.
	 *
	 * The observer function also returns a function which is called when the control scheme
	 * is changed (or the observation process is stopped). This is useful for disconnecting
	 * any inputs regarding the last-observed control scheme.
	 *
	 * ```ts
	 * let stopObserving = preferred.ObserveControlScheme((preferred) => {
	 * 	const bin = new Bin();
	 * 	if (preferred === "MouseKeyboard") {
	 * 		// User last used a mouse or keyboard input, thus listen to relevant devices:
	 * 		const keyboard = bin.Add(new Keyboard());
	 * 		keyboard.GetKeyDownSignal(Key.Space).Connect(() => {
	 * 			print("Space pressed");
	 * 		});
	 * 	}
	 * 	return () => {
	 * 		// Clean up any connections once the preferred scheme changes:
	 * 		bin.Destroy();
	 * 	};
	 * });
	 * ```
	 */
	public ObserveControlScheme(observer: (preferred: ControlScheme) => CleanupFunc): () => void {
		let cleanup: CleanupFunc;
		cleanup = observer(this.controlScheme);
		const disconnectChanged = this.controlSchemeChanged.Connect((controlScheme) => {
			cleanup?.();
			cleanup = observer(controlScheme);
		});
		let observing = true;
		const stopObserving = () => {
			if (!observing) return;
			observing = false;
			disconnectChanged();
			if (cleanup !== undefined) {
				cleanup();
				cleanup = undefined;
			}
		};
		this.bin.Add(stopObserving);
		return stopObserving;
	}

	/** Cleans up any connections/observers. */
	public Destroy() {
		this.bin.Destroy();
	}
}
